File: secure-skeleton/src/main/java/br/avcaliani/secureskeleton/user/services/impl/UserService.java
package br.avcaliani.secureskeleton.user.services.impl;

import br.avcaliani.secureskeleton.user.model.entities.User;
import br.avcaliani.secureskeleton.user.services.IUserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

@Service
public class UserService implements IUserService {

    private static final Set<User> USERS;

    @Autowired
    private BCryptPasswordEncoder encoder;

    static {
        USERS = new HashSet<>();
    }

    @Override
    public User save(User user) throws Exception {

        if (user == null || user.getUsername() == null || user.getPassword() == null)
            throw new Exception("Invalid user data!");

        user.setId(UUID.randomUUID());
        user.setPassword(encoder.encode(user.getPassword()));

        if (!USERS.add(user))
            throw new Exception("User already exists!");

        return user;
    }

}

File: kotlin-app/src/main/kotlin/br/avcaliani/kotlin/app/service/UserService.kt
package br.avcaliani.kotlin.app.service

import br.avcaliani.kotlin.app.model.User
import br.avcaliani.kotlin.app.repository.UserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.*

@Service
class UserService {

    @Autowired
    private lateinit var repository: UserRepository

    fun find(): List<User> = repository.findAll().toList()

    fun find(id: String): User = repository.find(id) ?: throw RuntimeException("User not found!")

    fun remove(id: String) = repository.deleteById(id)

    fun save(user: User): User {

        if (user.name.trim().isEmpty())
            throw RuntimeException("User 'email' must be defined!")

        if (user.email.trim().isEmpty())
            throw RuntimeException("User 'email' must be defined!")

        return repository.save(User(UUID.randomUUID().toString(), user.name, user.email))
    }

}